File: shared_resources/src/main/java/com/thewizrd/shared_resources/locationdata/here/HERELocationProvider.java
package com.thewizrd.shared_resources.locationdata.here;

import android.util.Log;

import com.ibm.icu.util.ULocale;
import com.thewizrd.shared_resources.SimpleLibrary;
import com.thewizrd.shared_resources.controls.LocationQueryViewModel;
import com.thewizrd.shared_resources.locationdata.LocationProviderImpl;
import com.thewizrd.shared_resources.utils.JSONParser;
import com.thewizrd.shared_resources.utils.Logger;
import com.thewizrd.shared_resources.utils.StringUtils;
import com.thewizrd.shared_resources.utils.WeatherException;
import com.thewizrd.shared_resources.utils.WeatherUtils;
import com.thewizrd.shared_resources.utils.here.HEREOAuthUtils;
import com.thewizrd.shared_resources.weatherdata.WeatherAPI;

import java.io.IOException;
import java.io.InputStream;
import java.net.URLEncoder;
import java.util.Collection;
import java.util.Collections;
import java.util.HashSet;
import java.util.Locale;

import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.Response;

public final class HERELocationProvider extends LocationProviderImpl {
    private static final String AUTOCOMPLETE_QUERY_URL = "https://autocomplete.geocoder.ls.hereapi.com/6.2/suggest.json?query=%s&language=%s&maxresults=10";
    private static final String GEOLOCATION_QUERY_URL = "https://reverse.geocoder.ls.hereapi.com/6.2/reversegeocode.json" +
            "?prox=%s,150&mode=retrieveAddresses&maxresults=1&additionaldata=Country2,true&gen=9&jsonattributes=1" +
            "&locationattributes=adminInfo,timeZone,-mapView,-mapReference&language=%s";
    private static final String GEOCODER_QUERY_URL = "https://geocoder.ls.hereapi.com/6.2/geocode.json" +
            "?locationid=%s&mode=retrieveAddresses&maxresults=1&additionaldata=Country2,true&gen=9&jsonattributes=1" +
            "&locationattributes=adminInfo,timeZone,-mapView,-mapReference&language=%s";

    @Override
    public String getLocationAPI() {
        return WeatherAPI.HERE;
    }

    @Override
    public boolean supportsLocale() {
        return true;
    }

    @Override
    public boolean isKeyRequired() {
        return false;
    }

    @Override
    public Collection<LocationQueryViewModel> getLocations(String ac_query, String weatherAPI) throws WeatherException {
        Collection<LocationQueryViewModel> locations = null;

        // Limit amount of results shown
        int maxResults = 10;

        ULocale uLocale = ULocale.forLocale(Locale.getDefault());
        String locale = localeToLangCode(uLocale.getLanguage(), uLocale.toLanguageTag());

        OkHttpClient client = SimpleLibrary.getInstance().getHttpClient();
        Response response = null;
        WeatherException wEx = null;

        try {
            final String authorization = HEREOAuthUtils.getBearerToken(false);

            if (StringUtils.isNullOrWhitespace(authorization)) {
                throw new WeatherException(WeatherUtils.ErrorStatus.NETWORKERROR);
            }

            Request request = new Request.Builder()
                    .url(String.format(AUTOCOMPLETE_QUERY_URL, URLEncoder.encode(ac_query, "UTF-8"), locale))
                    .addHeader("Authorization", authorization)
                    .build();

            // Connect to webstream
            response = client.newCall(request).execute();
            final InputStream stream = response.body().byteStream();

            // Load data
            locations = new HashSet<>(); // Use HashSet to avoid duplicate location (names)
            AutoCompleteQuery root = JSONParser.deserializer(stream, AutoCompleteQuery.class);

            for (SuggestionsItem result : root.getSuggestions()) {
                boolean added = false;
                // Filter: only store city results
                if ("city".equals(result.getMatchLevel())
                        || "district".equals(result.getMatchLevel())
                        || "postalCode".equals(result.getMatchLevel()))
                    added = locations.add(new LocationQueryViewModel(result, weatherAPI));
                else
                    continue;

                // Limit amount of results
                if (added) {
                    maxResults--;
                    if (maxResults <= 0)
                        break;
                }
            }

            // End Stream
            stream.close();
        } catch (Exception ex) {
            if (ex instanceof IOException) {
                wEx = new WeatherException(WeatherUtils.ErrorStatus.NETWORKERROR);
            }
            Logger.writeLine(Log.ERROR, ex, "HEREWeatherProvider: error getting locations");
        } finally {
            if (response != null)
                response.close();
        }

        if (wEx != null)
            throw wEx;

        if (locations == null || locations.size() == 0) {
            locations = Collections.singletonList(new LocationQueryViewModel());
        }

        return locations;
    }

    @Override
    public LocationQueryViewModel getLocation(WeatherUtils.Coordinate coord, String weatherAPI) throws WeatherException {
        LocationQueryViewModel location = null;

        String location_query = String.format(Locale.ROOT, "%s,%s", coord.getLatitude(), coord.getLongitude());

        ULocale uLocale = ULocale.forLocale(Locale.getDefault());
        String locale = localeToLangCode(uLocale.getLanguage(), uLocale.toLanguageTag());

        OkHttpClient client = SimpleLibrary.getInstance().getHttpClient();
        Response response = null;
        ResultItem result = null;
        WeatherException wEx = null;

        try {
            final String authorization = HEREOAuthUtils.getBearerToken(false);

            if (StringUtils.isNullOrWhitespace(authorization)) {
                throw new WeatherException(WeatherUtils.ErrorStatus.NETWORKERROR);
            }

            Request request = new Request.Builder()
                    .url(String.format(GEOLOCATION_QUERY_URL, location_query, locale))
                    .addHeader("Authorization", authorization)
                    .build();

            // Connect to webstream
            response = client.newCall(request).execute();
            final InputStream stream = response.body().byteStream();

            // Load data
            Geo_Rootobject root = JSONParser.deserializer(stream, Geo_Rootobject.class);

            if (root.getResponse().getView().size() > 0 && root.getResponse().getView().get(0).getResult().size() > 0)
                result = root.getResponse().getView().get(0).getResult().get(0);

            // End Stream
            stream.close();
        } catch (Exception ex) {
            result = null;
            if (ex instanceof IOException) {
                wEx = new WeatherException(WeatherUtils.ErrorStatus.NETWORKERROR);
            }
            Logger.writeLine(Log.ERROR, ex, "HEREWeatherProvider: error getting location");
        } finally {
            if (response != null)
                response.close();
        }

        if (wEx != null)
            throw wEx;

        if (result != null && !StringUtils.isNullOrWhitespace(result.getLocation().getLocationId()))
            location = new LocationQueryViewModel(result, weatherAPI);
        else
            location = new LocationQueryViewModel();

        return location;
    }

    public LocationQueryViewModel getLocationfromLocID(String locationID, String weatherAPI) throws WeatherException {
        LocationQueryViewModel location = null;

        ULocale uLocale = ULocale.forLocale(Locale.getDefault());
        String locale = localeToLangCode(uLocale.getLanguage(), uLocale.toLanguageTag());

        OkHttpClient client = SimpleLibrary.getInstance().getHttpClient();
        Response response = null;
        ResultItem result = null;
        WeatherException wEx = null;

        try {
            final String authorization = HEREOAuthUtils.getBearerToken(false);

            if (StringUtils.isNullOrWhitespace(authorization)) {
                throw new WeatherException(WeatherUtils.ErrorStatus.NETWORKERROR);
            }

            Request request = new Request.Builder()
                    .url(String.format(GEOCODER_QUERY_URL, locationID, locale))
                    .addHeader("Authorization", authorization)
                    .build();

            // Connect to webstream
            response = client.newCall(request).execute();
            final InputStream stream = response.body().byteStream();

            // Load data
            Geo_Rootobject root = JSONParser.deserializer(stream, Geo_Rootobject.class);

            if (root.getResponse().getView().size() > 0 && root.getResponse().getView().get(0).getResult().size() > 0)
                result = root.getResponse().getView().get(0).getResult().get(0);

            // End Stream
            stream.close();
        } catch (Exception ex) {
            result = null;
            if (ex instanceof IOException) {
                wEx = new WeatherException(WeatherUtils.ErrorStatus.NETWORKERROR);
            }
            Logger.writeLine(Log.ERROR, ex, "HEREWeatherProvider: error getting location");
        } finally {
            if (response != null)
                response.close();
        }

        if (wEx != null)
            throw wEx;

        if (result != null && !StringUtils.isNullOrWhitespace(result.getLocation().getLocationId()))
            location = new LocationQueryViewModel(result, weatherAPI);
        else
            location = new LocationQueryViewModel();

        return location;
    }

    @Override
    public boolean isKeyValid(String key) {
        return false;
    }

    @Override
    public String getAPIKey() {
        return null;
    }

    @Override
    public String localeToLangCode(String iso, String name) {
        return name;
    }

}

File: shared_resources/src/main/java/com/thewizrd/shared_resources/locationdata/here/HERELocationProvider.kt
package com.thewizrd.shared_resources.locationdata.here

import android.util.Log
import androidx.annotation.RestrictTo
import com.ibm.icu.util.ULocale
import com.thewizrd.shared_resources.SimpleLibrary
import com.thewizrd.shared_resources.controls.LocationQueryViewModel
import com.thewizrd.shared_resources.locationdata.LocationProviderImpl
import com.thewizrd.shared_resources.okhttp3.OkHttp3Utils.await
import com.thewizrd.shared_resources.okhttp3.OkHttp3Utils.getStream
import com.thewizrd.shared_resources.utils.*
import com.thewizrd.shared_resources.utils.here.HEREOAuthUtils
import com.thewizrd.shared_resources.weatherdata.WeatherAPI
import com.thewizrd.shared_resources.weatherdata.WeatherAPI.LocationProviders
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import okhttp3.CacheControl
import okhttp3.Request
import okhttp3.Response
import okhttp3.internal.closeQuietly
import java.io.IOException
import java.net.URLEncoder
import java.text.DecimalFormat
import java.util.*
import java.util.concurrent.TimeUnit

class HERELocationProvider  // Keep hidden for now
@RestrictTo(RestrictTo.Scope.TESTS) private constructor() : LocationProviderImpl() {
    companion object {
        private const val AUTOCOMPLETE_QUERY_URL = "https://autocomplete.geocoder.ls.hereapi.com/6.2/suggest.json?query=%s&language=%s&maxresults=10"
        private const val GEOLOCATION_QUERY_URL = "https://reverse.geocoder.ls.hereapi.com/6.2/reversegeocode.json" +
                                                  "?prox=%s,150&mode=retrieveAreas&maxresults=1&additionaldata=Country2,true&gen=9&jsonattributes=1" +
                                                  "&locationattributes=adminInfo,timeZone,-mapView,-mapReference&language=%s"
        private const val GEOCODER_QUERY_URL = "https://geocoder.ls.hereapi.com/6.2/geocode.json" +
                                               "?locationid=%s&mode=retrieveAreas&maxresults=1&additionaldata=Country2,true&gen=9&jsonattributes=1" +
                                               "&locationattributes=adminInfo,timeZone,-mapView,-mapReference&language=%s"
    }

    @LocationProviders
    override fun getLocationAPI(): String {
        return WeatherAPI.HERE
    }

    override fun supportsLocale(): Boolean {
        return true
    }

    override fun isKeyRequired(): Boolean {
        return false
    }

    override fun needsLocationFromID(): Boolean {
        return true
    }

    @Throws(WeatherException::class)
    override suspend fun getLocations(ac_query: String?, weatherAPI: String?
    ): Collection<LocationQueryViewModel> = withContext(Dispatchers.IO) {
        var locations: Collection<LocationQueryViewModel>? = null

        // Limit amount of results shown
        var maxResults = 10

        val uLocale = ULocale.forLocale(LocaleUtils.getLocale())
        val locale = localeToLangCode(uLocale.language, uLocale.toLanguageTag())

        val client = SimpleLibrary.instance.httpClient
        var response: Response? = null
        var wEx: WeatherException? = null

        try {
            val authorization = HEREOAuthUtils.getBearerToken(false)

            if (authorization.isNullOrBlank()) {
                throw WeatherException(ErrorStatus.NETWORKERROR)
            }

            val request = Request.Builder()
                    .cacheControl(CacheControl.Builder()
                            .maxAge(1, TimeUnit.DAYS)
                            .build())
                    .url(String.format(AUTOCOMPLETE_QUERY_URL, URLEncoder.encode(ac_query, "UTF-8"), locale))
                    .addHeader("Authorization", authorization)
                    .build()

            // Connect to webstream
            response = client.newCall(request).await()
            val stream = response.getStream()

            // Load data
            locations = HashSet() // Use HashSet to avoid duplicate location (names)
            val root = JSONParser.deserializer<AutoCompleteQuery>(stream, AutoCompleteQuery::class.java)

            for (result in root.suggestions) {
                var added = false
                // Filter: only store city results
                added = if ("city" == result.matchLevel
                            || "district" == result.matchLevel
                            || "postalCode" == result.matchLevel) {
                    locations.add(createLocationModel(result!!, weatherAPI!!))
                } else {
                    continue
                }

                // Limit amount of results
                if (added) {
                    maxResults--
                    if (maxResults <= 0) break
                }
            }

            // End Stream
            stream.closeQuietly()
        } catch (ex: Exception) {
            if (ex is IOException) {
                wEx = WeatherException(ErrorStatus.NETWORKERROR)
            }
            Logger.writeLine(Log.ERROR, ex, "HERELocationProvider: error getting locations")
        } finally {
            response?.closeQuietly()
        }

        if (wEx != null) throw wEx

        if (locations.isNullOrEmpty()) {
            locations = listOf(LocationQueryViewModel())
        }

        return@withContext locations
    }

    @Throws(WeatherException::class)
    override suspend fun getLocation(coordinate: Coordinate, weatherAPI: String?
    ): LocationQueryViewModel = withContext(Dispatchers.IO) {
        var location = super.getLocation(coordinate, weatherAPI)

        if (location != null) {
            return@withContext location
        }

        val df = DecimalFormat.getInstance(Locale.ROOT) as DecimalFormat
        df.applyPattern("0.####")

        val location_query = String.format(Locale.ROOT, "%s,%s", df.format(coordinate.latitude), df.format(coordinate.longitude))

        val uLocale = ULocale.forLocale(LocaleUtils.getLocale())
        val locale = localeToLangCode(uLocale.language, uLocale.toLanguageTag())

        val client = SimpleLibrary.instance.httpClient
        var response: Response? = null
        var result: ResultItem? = null
        var wEx: WeatherException? = null

        try {
            val authorization = HEREOAuthUtils.getBearerToken(false)

            if (authorization.isNullOrBlank()) {
                throw WeatherException(ErrorStatus.NETWORKERROR)
            }

            val request = Request.Builder()
                    .cacheControl(CacheControl.Builder()
                            .maxAge(1, TimeUnit.DAYS)
                            .build())
                    .url(String.format(GEOLOCATION_QUERY_URL, location_query, locale))
                    .addHeader("Authorization", authorization)
                    .build()

            // Connect to webstream
            response = client.newCall(request).await()
            val stream = response.getStream()

            // Load data
            val root = JSONParser.deserializer<Geo_Rootobject>(stream, Geo_Rootobject::class.java)

            if (!root.response.view.isNullOrEmpty() && !root.response.view[0]?.result.isNullOrEmpty())
                result = root.response.view[0].result[0]

            // End Stream
            stream.closeQuietly()
        } catch (ex: Exception) {
            result = null
            if (ex is IOException) {
                wEx = WeatherException(ErrorStatus.NETWORKERROR)
            }
            Logger.writeLine(Log.ERROR, ex, "HERELocationProvider: error getting location")
        } finally {
            response?.closeQuietly()
        }

        if (wEx != null) throw wEx

        location = if (!result?.location?.locationId.isNullOrBlank())
            createLocationModel(result!!, weatherAPI!!)
        else
            LocationQueryViewModel()

        return@withContext location
    }

    @Throws(WeatherException::class)
    override suspend fun getLocationFromID(model: LocationQueryViewModel): LocationQueryViewModel =
            withContext(Dispatchers.IO) {
                val uLocale = ULocale.forLocale(LocaleUtils.getLocale())
                val locale = localeToLangCode(uLocale.language, uLocale.toLanguageTag())

                val client = SimpleLibrary.instance.httpClient
                var response: Response? = null
                var result: ResultItem? = null
                var wEx: WeatherException? = null

                try {
                    val authorization = HEREOAuthUtils.getBearerToken(false)

                    if (authorization.isNullOrBlank()) {
                        throw WeatherException(ErrorStatus.NETWORKERROR)
                    }

                    val request = Request.Builder()
                            .cacheControl(CacheControl.Builder()
                                    .maxAge(1, TimeUnit.DAYS)
                                    .build())
                            .url(String.format(GEOCODER_QUERY_URL, model.locationQuery, locale))
                            .addHeader("Authorization", authorization)
                            .build()

                    // Connect to webstream
                    response = client.newCall(request).await()
                    val stream = response.getStream()

                    // Load data
                    val root = JSONParser.deserializer<Geo_Rootobject>(stream, Geo_Rootobject::class.java)

                    if (!root.response.view.isNullOrEmpty() && !root.response.view[0]?.result.isNullOrEmpty())
                        result = root.response.view[0].result[0]

                    // End Stream
                    stream.closeQuietly()
                } catch (ex: Exception) {
                    result = null
                    if (ex is IOException) {
                        wEx = WeatherException(ErrorStatus.NETWORKERROR)
                    }
                    Logger.writeLine(Log.ERROR, ex, "HERELocationProvider: error getting location")
                } finally {
                    response?.closeQuietly()
                }

                if (wEx != null) throw wEx

                return@withContext if (!result?.location?.locationId.isNullOrBlank())
                    createLocationModel(result!!, model.weatherSource!!)
                else
                    LocationQueryViewModel()
            }

    @Throws(WeatherException::class)
    override suspend fun getLocationFromName(model: LocationQueryViewModel
    ): LocationQueryViewModel? {
        return null
    }

    override suspend fun isKeyValid(key: String?): Boolean {
        return false
    }

    override fun getAPIKey(): String? {
        return null
    }

    override fun localeToLangCode(iso: String, name: String): String {
        return name
    }
}